pkg/arvo/lib/lm.hoon

```diff
@@ -0,0 +1,702 @@
+  ::
+  ::    Identity
+  ++  id
+    ~/  %id
+    |=  [m=@ud n=@ud]  ^-  @lms
+    ~_  leaf+"lagoon-fail"
+    =/  u  (zeros m n)
+    =/  i  1
+    |-
+      ?:  |(=(i m) =(i n))  u
+    $(i +(i), u (set u i i .1))  :: TODO test Hoon version (jet works)
+  ::
+  ::    Shape of matrix
+  ++  shape
+    |=  u=@lms  ^-  (list @ud)
+    (decantor (snag 0 (flop (rip 5 u))))
+  ::
+  ::    Produce a matrix from `(list (list @rs))`
+  ::    Rows across, columns "down" (meaning modulus m)
+  ++  not-rs
+    |=  a=@rs
+    ~_  leaf+"lagoon-fail"
+    (not 5 1 a)
+  ++  make
+    |=  [u=(list (list @rs))]  ^-  @lms
+    ~_  leaf+"lagoon-fail"
+    =/  m  `@ud`(lent u)
+    =/  n  `@ud`(lent (snag 0 u))
+    =/  size  (mul m n)
+    =/  p  (cantor m n)
+    ~&  [m n size p]
+    =/  i  0  :: index over rows
+    =/  not-w  `(list @rs)`~
+    |-  ^-  @lms
+      ?:  =(i m)  `@lms`(not 5 +(size) (rep 5 (weld `(list @rs)`not-w ~[(not 5 1 p)])))
+      =/  not-u  `(list @rs)`(flop (turn (snag i u) not-rs))
+    $(i +(i), not-w `(list @rs)`(weld `(list @rs)`not-u `(list @rs)`not-w))
+  ++  unmake
+    |=  [u=@lms]  ^-  (list (list @rs))
+    ~_  leaf+"lagoon-fail"
+    %-  flop
+    =/  mn  `(list @ud)`(shape u)
+    =/  m   `@ud`-:mn
+    =/  n   `@ud`+<:mn
+    =/  size  (mul m n)
+    =/  p  (cantor m n)
+    =/  nan  0xffff.ffff
+    =/  ut  (swp 5 (cat 5 nan (cat 5 u nan)))
+    =/  rc  `(list @rs)`(oust [size 1] (oust [0 2] (rip 5 ut)))
+    =/  q  `(list (list @rs))`~
+    =/  i  0  :: index over rows
+    |-  ^-  (list (list @rs))
+      ?:  =(i m)  `(list (list @rs))`q
+      =/  r  (mul i n)
+      =/  v  `(list @rs)`(swag [r n] `(list @rs)`rc)
+    $(i +(i), q `(list (list @rs))`(weld `(list (list @rs))`~[v] `(list (list @rs))`q))
+  ::
+  ::    Pretty-print the contents of the matrix.
+  ++  pprint  !!
+  ::
+  ::    Get the value at an index, using mathematical indices 1..n.
+  ++  get
+    ~/  %get
+    |=  [u=@lms i=@ud j=@ud]  ^-  @rs
+    ~_  leaf+"lagoon-fail"
+    =/  mn  `(list @ud)`(shape u)
+    =/  m   `@ud`-:mn
+    =/  n   `@ud`+<:mn
+    =/  row  (snag (dec i) (unmake u))
+    (snag (dec j) row)
+  ::
+  ::    Set the value of an element within a matrix, using math indices 1..n.
+  ++  set
+    ~/  %set
+    |=  [u=@lms i=@ud j=@ud x=@rs]  ^-  @lms
+    ~_  leaf+"lagoon-fail"
+    `@lms`(setr u i (set:lvs (getr u i) j x))
+  ::
+  ::    Get the value of a column as @lvs in 1..n
+  ++  getc
+    ~/  %getc
+    |=  [u=@lms i=@ud]  ^-  @lvs
+    ~_  leaf+"lagoon-fail"
+    =/  mn  `(list @ud)`(shape u)
+    =/  m   `@ud`-:mn
+    =/  n   `@ud`+<:mn
+    =/  w  `(list @rs)`~
+    =/  uu  `(list (list @rs))`(unmake u)
+    =/  count  1
+    |-  ^-  @lvs
+      ?:  (gth count m)  `@lvs`(make:lvs w)
+      =/  row  `(list @rs)`(snag (dec count) `(list (list @rs))`uu)
+      =/  elem  `@rs`(snag (dec i) row)
+    $(count +(count), w `(list @rs)`(weld w `(list @rs)`~[elem]))
```

Comment:
I feel like a `makec/unmakec/maker/unmaker` that go <-> `(list @lv)` would help a lot with these